docs/building-extension-modules.md

```diff
@@ -0,0 +1,338 @@
+---
+title: "Building extension modules"
+draft: false
+type: docs
+layout: single
+
+menu:
+  docs:
+    weight: 125
+---
+
+# Building Extension Modules
+
+{{% warning %}}
+While this feature has been around since almost the beginning of the Poetry project and has needed minimal changes,
+it is still considered unstable. You can participate in the discussions about stabilizing this feature
+[here](https://github.com/python-poetry/poetry/issues/2740).
+
+And as always, your contributions towards the goal of improving this feature are also welcome.
+{{% /warning %}}
+
+Poetry allows a project developer to introduce support for, build and distribute native extensions within their project.
+In order to achieve this, at the highest level, the following steps are required.
+
+{{< steps >}}
+{{< step >}}
+**Add Build Dependencies**
+
+The build dependencies, in this context, refer to those Python packages that required in order to successfully execute
+your build script. Common examples include `cython`, `meson`, `maturin`, `setuptools` etc., depending on how your
+extension is built.
+
+{{% note %}}
+You must assume that only python built-ins are available by default in a build environment. This means, if you use need
+even packages like `setuptools`, it must be explicitly declared.
+{{% /note %}}
+
+The necessary build dependencies must be added to the `build-system.requires` section of your `pyproject.toml` file.
+
+```toml
+[build-system]
+requires = ["poetry-core", "setuptools", "cython"]
+build-backend = "poetry.core.masonry.api"
+```
+
+{{% note %}}
+If you wish to develop the build script within your project's virtual environment, then you must also add the
+dependencies to your project explicitly to a dependency group - the name of which is not important.
+
+```sh
+poetry add --group=build setuptools cython
+```
+{{% /note %}}
+
+{{< /step >}}
+
+{{< step >}}
+**Add Build Script**
+
+The build script can be free-form Python script that uses any dependency specified in the previous step. This can be
+named as needed, but **must** be located within the project root directory (or a subdirectory) and also **must**
+be included in your source distribution. You can see the [example snippets section]({{< relref "#example-snippets" >}})
+for inspiration.
+
+```toml
+[tool.poetry.build]
+script = "build-extension.py"
+```
+
+{{% note %}}
+The name of the build script is arbitrary. Common practice has been to name it `build.py`, however this is not
+mandatory. You **should** consider [using a subdirectory]({{< relref "#can-i-store-the-build-script-in-a-subdirectory" >}})
+if feasible.
+{{% /note %}}
+
+{{< /step >}}
+
+{{< step >}}
+**Specify Distribution Files**
+
+{{% warning %}}
+The following is an example, and should not be considered as complete.
+{{% /warning %}}
+
+```toml
+[tool.poetry]
+...
+packages = [
+    { include = "package", from = "src" }
+]
+include = [
+    { path = "src/package/**/*.so", format = "wheel" },
```

Comment:
Does this work? Or is the `*.so` put into a `src` folder into the wheel? I remember such issues because we do not support `from` and `to` here as we do for `packages`.

Reply:
You are probably correct. I'll see if I can fix that if it's the case.

Reply:
Yeah, the current include/exclude stuff is a bit of a mess. I have simplified the doc to work with what we have right now.

Reply:
The following seems to do the right things for now.

```
{ path = "src/package/**/*.so", format = "wheel" },
```

Reply:
```console
$ unzip *.whl
Archive:  poetry_cython_example-0.1.0-cp313-cp313-manylinux_2_40_x86_64.whl
  inflating: package/__init__.py     
  inflating: package/example.cpython-313-x86_64-linux-gnu.so  
  inflating: package/example.pyx     
  inflating: poetry_cython_example-0.1.0.dist-info/METADATA  
  inflating: poetry_cython_example-0.1.0.dist-info/WHEEL  
  inflating: poetry_cython_example-0.1.0.dist-info/RECORD 
```

Reply:
I debugged this and it only works because the `*.so` file has already been picked up by the package include (with the correct target directory) so that the other include (with the wrong target directory) is a NOP (because we do not include files twice). Otherwise, it would end in the wrong directory.

In other words, you can remove the `include` and get the same result.

Reply:
Well it does ignore the c files being generated though, so this means the package directory is not necessarily all included. Or are you saying `*.so` files are treated differently?

Reply:
Ah, I did not consider that the `*.so` is in your `.gitignore`. That makes the difference. This is really weird: The `*.so` file would be ignored without the `include` because of `.gitignore`. However, the `include` only ensures that the `*.so` is not ignored but it is not used itself, but the `*.so` is picked up from the package include. Thus, the example makes sense, especially with `.gitignore`.

Reply:
I guess we should look at the include/exclude stuff at some point.